Prevent fd leaks in ExecutingCommand

shvo123 opened the pull request:
in Solaris11 OS, pipes left open in case their corresponding OutputStream - stdout, InputStream - stdin, ErrorStream - stderr are not explicitly closed.

shvo123 commented:
![patch](https://user-images.githubusercontent.com/10284452/142759639-069db07a-981f-45dd-b715-7484fc5e97fc.jpg)

mprins commented:
Does this fix an actual problem? Because `getProcessOutputAndDestroy(` doesn't actually open all those output streams... also I think it would be better to fix there

shvo123 commented:
it is really required. in Solaris11 OS it left open fds until eventually it
has "too many open files"


On Sun, Nov 21, 2021 at 1:22 PM Mark Prins ***@***.***> wrote:

> Does this fix an actual problem? Because getProcessOutputAndDestroy(
> doesn't actually open all those output streams... also I think it would be
> better to fix there
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/oshi/oshi/pull/1781#issuecomment-974797734>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOO3JHQFJIDGECZFNZXJHDUNDI6VANCNFSM5IPAHBVQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Rgds,
David Herschler Shvo

shvo123 commented:
The finally I added serves the possibility that  p = Runtime.getRuntime().exec(cmdToRunWithArgs, envp); failed.
To be on the safe side, I added as you suggested the streams closing prior to the destroy in getProcessOutputAndDestroy as well.10x,David

dbwiddis commented:
This seems like overkill, and I think it's unnecessary:
 - If the `exec()` call fails with an exception, there is no process object to return; `p` is null, and the finally block you've added does nothing.
 - If `exec()` returns successfully, `getProcessOutputAndDestroy()` has a `finally` block which destroys the process. Java closes all process streams when the process is destroyed, so I believe that the fix I did in #1767 already handles this issue.

If you want to move the `finally` block from `getProcessOutputAndDestroy()` to this method (and change the name), that seems a simpler approach and I wouldn't complain.

dbwiddis commented on oshi-core/src/main/java/oshi/util/ExecutingCommand.java:
```diff
@@ -116,11 +116,33 @@ private ExecutingCommand() {
         } catch (SecurityException | IOException e) {
             LOG.trace("Couldn't run command {}: {}", Arrays.toString(cmdToRunWithArgs), e.getMessage());
+        } finally {
+            if (p != null) {
```
If `exec()` fails with an exception, we will never execute this block.

shvo123 replied:
Hi Daniel,
You wrote "if you want to move the finally block from getProcessOutputAndDestroy() to this method (and change the name)...".
I did not understand what method and what name you are referring to.
10x,
David

dbwiddis replied:
> I did not understand what method and what name you are referring to.

A few weeks ago, I noted the case you're referring to in this PR, where `exec()` fails.  The PR #1767 was intended to address this case.  My solution in that case was to add a `finally` block in the method `getProcessOutputAndDestroy()` which terminated the process (and closed all its resources).

If you think the `p.destroy()` call is better placed in a `finally` block in the `runNative()` method you've addressed in your PR, I'm okay with that.  The point I'm making is that the `destroy()` method accomplishes all the stream closures that you are proposing in this PR.

Regarding the "change the name" I intentionally added "AndDestroy" to the name to make it clear that this method call would kill the process after collecting its output.  If you're moving the `destroy()` call out of that method then its name should just be `getProcessOutput()`.

dbwiddis commented on oshi-core/src/main/java/oshi/util/ExecutingCommand.java:
```diff
@@ -116,11 +116,33 @@ private ExecutingCommand() {
+            if (p != null) {
+                if (p.getOutputStream() != null) {
+                    try {
+                        p.getOutputStream().close();
```
If `exec()` succeeded, `p` is destroyed and it has no output stream to close.

shvo123 replied:
This is not true, In Solaris after succesfull execution of exec() I can see open pipes that are opened. After I added this enhancement problem solved. I am calling OSHI every 5 minutes and i can see the increase y using ls -al /proc/${PID}/fd. the p.destory() does not close these pipes. did you realy tried it on Solaris? I can show it to you in our Solaris.

dbwiddis replied:
I'd be happy to review the logs and any information you have

1. Are you using the latest OSHI _**snapshot**_ version (including the #1767 fix referenced)?  If not, I acknowledge the bug in the latest _released_ version but think it's been fixed by #1767.
2. Please let me know exactly what JVM you are using as OpenJDK has dropped Solaris support so I do not know whether your port properly handles this case.

dbwiddis commented:
In the case `exec()` fails:
 - `Runtime.exec()` calls [`ProcessBuilder.start()`](https://github.com/openjdk/jdk/blob/7e73bca0b7a34af9fb73780491951539815651b4/src/java.base/share/classes/java/lang/Runtime.java#L615)
 - `ProcessBuilder.start()` calls [`ProcessImpl.start()`](https://github.com/openjdk/jdk/blob/739769c8fc4b496f08a92225a12d07414537b6c0/src/java.base/share/classes/java/lang/ProcessBuilder.java#L1108)
 - `ProcessImpl.start()` obtains the native file descriptors [inside a `try` block](https://github.com/openjdk/jdk/blob/739769c8fc4b496f08a92225a12d07414537b6c0/src/java.base/unix/classes/java/lang/ProcessImpl.java#L194), and then creates copies right before returning the new (started) process object, before closing the orignal native file descriptors in [a `finally` block](https://github.com/openjdk/jdk/blob/739769c8fc4b496f08a92225a12d07414537b6c0/src/java.base/unix/classes/java/lang/ProcessImpl.java#L264).

Takeaway: the original native fds are closed, and the only way the streams attached to p will be useful is if `p` is successfully returned.

In the case `exec()` succeeds, and `p` is returned non-null with copies of the fds, here's the [`destroy()` implementation](https://github.com/openjdk/jdk/blob/6a34176e9ce8bea99c061eda7a4844e856ae1cbc/jdk/src/java.base/unix/classes/java/lang/UNIXProcess.java#L432) for Solaris (last version before Solaris support was dropped). It includes an attempt to close the stdin, stdout, and stderr streams.
```java
try {
    stdin.close();
    if (stdout_inner_stream != null)
        stdout_inner_stream.closeDeferred(stdout);
    if (stderr instanceof DeferredCloseInputStream)
        ((DeferredCloseInputStream) stderr)
            .closeDeferred(stderr);
} catch (IOException e) {
    // ignore
}
```

shvo123 commented:
Dear Daniel,
Ohh.. now I understand you. and I would submit the change as you meant.

I saw parsing errors in both Solaris and AIX see attached, Let me know
whether , You want me to open issues for them.
It seems to me that the line read from ps output contains binary zeroes.

I can handle it, provided that it will be released next minor version.



10x,
David


On Mon, Nov 22, 2021 at 9:17 AM Daniel Widdis ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In oshi-core/src/main/java/oshi/util/ExecutingCommand.java
> <https://github.com/oshi/oshi/pull/1781#discussion_r754001479>:
>
> >              return getProcessOutputAndDestroy(p, cmdToRunWithArgs);
>          } catch (SecurityException | IOException e) {
>              LOG.trace("Couldn't run command {}: {}", Arrays.toString(cmdToRunWithArgs), e.getMessage());
> +        } finally {
> +            if (p != null) {
>
> I did not understand what method and what name you are referring to.
>
> A few weeks ago, I noted the case you're referring to in this PR, where
> exec() fails. The PR #1767 <https://github.com/oshi/oshi/pull/1767> was
> intended to address this case. My solution in that case was to add a
> finally block in the method getProcessOutputAndDestroy() which terminated
> the process (and closed all its resources).
>
> If you think the p.destroy() call is better placed in a finally block in
> the runNative() method you've addressed in your PR, I'm okay with that.
> The point I'm making is that the destroy() method accomplishes all the
> stream closures that you are proposing in this PR.
>
> Regarding the "change the name" I intentionally added "AndDestroy" to the
> name to make it clear that this method call would kill the process after
> collecting its output. If you're moving the destroy() call out of that
> method then its name should just be getProcessOutput().
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/oshi/oshi/pull/1781#discussion_r754001479>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOO3JFT4F6NR5OBT5VI4FLUNHU7VANCNFSM5IPAHBVQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Rgds,
David Herschler Shvo

uname -a
SunOS ps-tlv-t84 5.11 11.4.11.4.0 sun4v sparc sun4v
~/bmcjava/bmcjava-V4/bin/java -version
openjdk version "11.0.10" 2021-01-19 LTS
OpenJDK Runtime Environment (build 11.0.10+9-LTS)
OpenJDK 64-Bit Server VM (build 11.0.10+9-LTS, mixed mode)
errors:
1123 11:45:00.354 TRACE MeasureAllComponents1  oshi.util.ParseUtil                      - ***@***.*** 0.null didn't parse. Returning default. {}
java.lang.NumberFormatException: For input string: "0.null"
	at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2054)
	at java.base/jdk.internal.math.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.base/java.lang.Double.parseDouble(Double.java:543)
	at oshi.util.ParseUtil.parseDoubleOrDefault(ParseUtil.java:547)
	at oshi.util.ParseUtil.parseDHMSOrDefault(ParseUtil.java:578)
	at oshi.software.os.unix.solaris.SolarisOSThread.updateAttributes(SolarisOSThread.java:139)
	at oshi.software.os.unix.solaris.SolarisOSThread.<init>(SolarisOSThread.java:56)
	at oshi.software.os.unix.solaris.SolarisOSProcess.getThreadDetails(SolarisOSProcess.java:305)
	at com.bmc.ctm.agent.metric.ProcInfo.<init>(ProcInfo.java:82)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.scheduleProcess_aroundBody0(MeasureAllComponents.java:199)
	at com.bmc.ctm.agent.metric.MeasureAllComponents$AjcClosure1.run(MeasureAllComponents.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:167)
	at com.bmc.ctm.common.logger.LoggerContextAspect.storeContextMethod(LoggerContextAspect.java:65)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.scheduleProcess(MeasureAllComponents.java:183)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.lambda$0(MeasureAllComponents.java:315)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:95)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
1123 11:45:00.354 TRACE MeasureAllComponents1  oshi.util.ParseUtil                      - ***@***.*** 0.null didn't parse. Returning default. {}
java.lang.NumberFormatException: For input string: "0.null"
	at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2054)
	at java.base/jdk.internal.math.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.base/java.lang.Double.parseDouble(Double.java:543)
	at oshi.util.ParseUtil.parseDoubleOrDefault(ParseUtil.java:547)
	at oshi.util.ParseUtil.parseDHMSOrDefault(ParseUtil.java:578)
	at oshi.software.os.unix.solaris.SolarisOSThread.updateAttributes(SolarisOSThread.java:139)
	at oshi.software.os.unix.solaris.SolarisOSThread.<init>(SolarisOSThread.java:56)
	at oshi.software.os.unix.solaris.SolarisOSProcess.getThreadDetails(SolarisOSProcess.java:305)
	at com.bmc.ctm.agent.metric.ProcInfo.<init>(ProcInfo.java:82)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.scheduleProcess_aroundBody0(MeasureAllComponents.java:199)
	at com.bmc.ctm.agent.metric.MeasureAllComponents$AjcClosure1.run(MeasureAllComponents.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:167)
	at com.bmc.ctm.common.logger.LoggerContextAspect.storeContextMethod(LoggerContextAspect.java:65)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.scheduleProcess(MeasureAllComponents.java:183)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.lambda$0(MeasureAllComponents.java:315)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:95)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
1123 11:45:00.354 TRACE MeasureAllComponents1  oshi.util.ParseUtil                      - ***@***.*** 0.null didn't parse. Returning default. {}
java.lang.NumberFormatException: For input string: "0.null"
	at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2054)
	at java.base/jdk.internal.math.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.base/java.lang.Double.parseDouble(Double.java:543)
	at oshi.util.ParseUtil.parseDoubleOrDefault(ParseUtil.java:547)
	at oshi.util.ParseUtil.parseDHMSOrDefault(ParseUtil.java:578)
	at oshi.software.os.unix.solaris.SolarisOSThread.updateAttributes(SolarisOSThread.java:144)
	at oshi.software.os.unix.solaris.SolarisOSThread.<init>(SolarisOSThread.java:56)
	at oshi.software.os.unix.solaris.SolarisOSProcess.getThreadDetails(SolarisOSProcess.java:305)
	at com.bmc.ctm.agent.metric.ProcInfo.<init>(ProcInfo.java:82)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.scheduleProcess_aroundBody0(MeasureAllComponents.java:199)
	at com.bmc.ctm.agent.metric.MeasureAllComponents$AjcClosure1.run(MeasureAllComponents.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:167)
	at com.bmc.ctm.common.logger.LoggerContextAspect.storeContextMethod(LoggerContextAspect.java:65)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.scheduleProcess(MeasureAllComponents.java:183)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.lambda$0(MeasureAllComponents.java:315)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:95)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
AIX va-tlv-ctm-qa23 2 7 00C81D104B00
AIX version 7.2
~/bmcjava/bmcjava-V4/bin/java -version
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.10+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.10+9, mixed mode)
errors:
1123 04:35:01.091 TRACE MeasureAllComponents1  oshi.util.ParseUtil                      - ***@***.*** 0.null didn't parse. Returning default. {}
java.lang.NumberFormatException: For input string: "0.null"
	at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(Unknown Source)
	at java.base/jdk.internal.math.FloatingDecimal.parseDouble(Unknown Source)
	at java.base/java.lang.Double.parseDouble(Unknown Source)
	at oshi.util.ParseUtil.parseDoubleOrDefault(ParseUtil.java:547)
	at oshi.util.ParseUtil.parseDHMSOrDefault(ParseUtil.java:578)
	at oshi.software.os.unix.aix.AixOSProcess.updateAttributes(AixOSProcess.java:335)
	at oshi.software.os.unix.aix.AixOSProcess.<init>(AixOSProcess.java:101)
	at oshi.software.os.unix.aix.AixOperatingSystem.getProcessListFromPS(AixOperatingSystem.java:181)
	at oshi.software.os.unix.aix.AixOperatingSystem.getProcess(AixOperatingSystem.java:155)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline.collect(Unknown Source)
	at oshi.software.os.OperatingSystem.getProcesses(OperatingSystem.java:217)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.scheduleProcess_aroundBody0(MeasureAllComponents.java:197)
	at com.bmc.ctm.agent.metric.MeasureAllComponents$AjcClosure1.run(MeasureAllComponents.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:167)
	at com.bmc.ctm.common.logger.LoggerContextAspect.storeContextMethod(LoggerContextAspect.java:65)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.scheduleProcess(MeasureAllComponents.java:183)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.lambda$0(MeasureAllComponents.java:315)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:95)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
1123 04:35:01.091 TRACE MeasureAllComponents1  oshi.util.ParseUtil                      - ***@***.*** 0.null didn't parse. Returning default. {}
java.lang.NumberFormatException: For input string: "0.null"
	at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(Unknown Source)
	at java.base/jdk.internal.math.FloatingDecimal.parseDouble(Unknown Source)
	at java.base/java.lang.Double.parseDouble(Unknown Source)
	at oshi.util.ParseUtil.parseDoubleOrDefault(ParseUtil.java:547)
	at oshi.util.ParseUtil.parseDHMSOrDefault(ParseUtil.java:578)
	at oshi.software.os.unix.aix.AixOSProcess.updateAttributes(AixOSProcess.java:335)
	at oshi.software.os.unix.aix.AixOSProcess.<init>(AixOSProcess.java:101)
	at oshi.software.os.unix.aix.AixOperatingSystem.getProcessListFromPS(AixOperatingSystem.java:181)
	at oshi.software.os.unix.aix.AixOperatingSystem.getProcess(AixOperatingSystem.java:155)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline.collect(Unknown Source)
	at oshi.software.os.OperatingSystem.getProcesses(OperatingSystem.java:217)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.scheduleProcess_aroundBody0(MeasureAllComponents.java:197)
	at com.bmc.ctm.agent.metric.MeasureAllComponents$AjcClosure1.run(MeasureAllComponents.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:167)
	at com.bmc.ctm.common.logger.LoggerContextAspect.storeContextMethod(LoggerContextAspect.java:65)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.scheduleProcess(MeasureAllComponents.java:183)
	at com.bmc.ctm.agent.metric.MeasureAllComponents.lambda$0(MeasureAllComponents.java:315)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:95)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
1123 04:35:01.590 TRACE MeasureAllComponents1  oshi.util.ExecutingCommand               - ***@***.*** Couldn't run command [pflags, 26607956]: Cannot run program "pflags": error=2, A file or directory in the path name does not exist.

dbwiddis commented:
Thanks for letting me know about the parsing problems. Can you please open a new issue for that?  If possible to include the actual `ps` output causing the problems that would be very helpful.  It's hard to tell from the output exactly where the error is.

I also notice that you have access to an AIX box.  We haven't had (and still don't have) the ability to run any tests on AIX since March of this year because the AIX server we'd been using at PolharHome has been down since then.  Even worse, JNA (our primary dependency) was unable to build new AIX binaries for the 5.10.0 release.  Is there any way you can assist us (and JNA) with access to an AIX server for some of this testing/building?

dbwiddis commented:
Disregard the parsing error. I found it and added issue #1788.  The question about AIX server access still stands! :)

shvo123 commented:
parseUtil.java line 578
I changed it from:
 milliseconds += (long) (1000 * parseDoubleOrDefault("0." + m.group(5),
0d));

            milliseconds += (long) (1000 * parseDoubleOrDefault("0." +
(m.group(5) != null ? m.group(5) : "0"), 0d));

On Tue, Nov 23, 2021 at 11:39 PM Daniel Widdis ***@***.***>
wrote:

> Disregard the parsing error. I found it and added issue #1788
> <https://github.com/oshi/oshi/issues/1788>. The question about AIX server
> access still stands! :)
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/oshi/oshi/pull/1781#issuecomment-977191284>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOO3JFZH3C4CA6U4NZNKD3UNQC27ANCNFSM5IPAHBVQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Rgds,
David Herschler Shvo

dbwiddis commented:
Did you mean to close this? :)

Also while your fix works you're still spending processing time parsing a constant string "0".  Better just to completely skip the whole += bit.

Or if you want the ternary operator, do `milliseconds += (m.group(5) == null) ? 0 : parseDouble...`

shvo123 commented:
Dear Daniel,
AIX 7.2
> ps -o
st,pid,ppid,user,uid,group,gid,thcount,pri,vsize,rssize,etime,time,comm,pagein,args
ST      PID     PPID     USER UID    GROUP   GID THCNT PRI   VSZ   RSS
ELAPSED        TIME COMMAND  PAGEIN COMMAND
A  64881208 37421480 controlm 210      dba   202     1  60   704  1052
  00:16    00:00:00 tcsh          0 -tcsh
A  28968140 64881208 controlm 210   system     0     1  61  2092  2212
  00:00    00:00:00 ps            0 ps -o
st,pid,ppid,user,uid,group,gid,thcount,pri,vsize,rssize,etime,time,com
>

AIX 5.3
ps -o
st,pid,ppid,user,uid,group,gid,thcount,pri,vsize,rssize,etime,time,comm,pagein,args
ST    PID   PPID     USER    UID    GROUP   GID THCNT PRI   VSZ   RSS
ELAPSED        TIME COMMAND  PAGEIN COMMAND
A  397454 794624 dherschl 123860      mpm   313     1  60   876  1220
02:09    00:00:00 tcsh          1 -tcsh
A  790554 397454 dherschl 123860   system     0     1  60   768   860
00:00    00:00:00 ps            0 ps -o
st,pid,ppid,user,uid,group,gid,thcount,pri,vsize,rssize,etime,time,comm
>

Solaris
ps -o s,pid,ppid,user,uid,group,gid,nlwp,pri,vsz,rss,etime,time,comm,args
S   PID  PPID     USER   UID    GROUP   GID NLWP PRI  VSZ      RSS
ELAPSED        TIME COMMAND
                         COMMAND
O  3873 50626    dshvo   100    staff    10    1  59 10160     6640
00:00       00:00 ps
                        ps -o
s,pid,ppid,user,uid,group,gid,nlwp,pri,vsz,rss,etime,time,comm,args
S 50626 50622    dshvo   100    staff    10    1  49 12360     8648
 20:19:40       00:00 -bash
                           -bash
S 55932 55930    dshvo   100    staff    10    1  59 22392    11264
 13:45:45       00:02 /export/home/dshvo/ctm/exe/p_ctmatw
                           /export/home/dshvo/ctm/exe/p_ctmatw -ATW_NAME
ATW000
S 55869     1    dshvo   100    staff    10    1  59 22392    11520
 13:45:46       00:04 /export/home/dshvo/ctm/exe/p_ctmag
                            /export/home/dshvo/ctm/exe/p_ctmag
S 55829     1    dshvo   100    staff    10  276  59 1077200   559904
 13:45:55       15:05 /export/home/dshvo/bmcjava/bmcjava-V4//bin/java
                           /export/home/dshvo/bmcjava/bmcjava-V4//bin/java
-Xmx256m -XX:+UseG1GC -XX:+CrashOnOutOfMemoryError -Djava.io.tmpdir=/tmp
-Djava.net.preferIPv4Stack=true -Doverride.default.services=
-Dspring.profiles.active=tcp -DCTMAG.CONFIG.DBGLVL=5
-Dctm.logs.dir=/export/home/dshvo/ctm/proclog
-Dlogging.config=/export/home/dshvo/ctm/data/logback.xml
-Dctm.data.dir=/export/home/dshvo/ctm/data -XX:+PrintFlagsFinal
-Dstdout=/export/home/dshvo/ctm/proclog/agjstd_55520-2021-11-23.0.tmp
-DLIMIT_LOG_VERSIONS=8 -DLIMIT_LOG_FILE_SIZE=10 -jar
/export/home/dshvo/ctm/exe/ag-app.jar
S 55930     1    dshvo   100    staff    10    1  59 22392    11264
 13:45:45       00:06 /export/home/dshvo/ctm/exe/p_ctmat
                            /export/home/dshvo/ctm/exe/p_ctmat

About helping you with AIX, our build machine is AIX desert 3 5
00CA277F4C00 means V 5.3

I would talk with my boss and I would recommend (of course) to cooperate.

10x,
David


On Tue, Nov 23, 2021 at 9:32 PM Daniel Widdis ***@***.***>
wrote:

> Thanks for letting me know about the parsing problems. Can you please open
> a new issue for that? If possible to include the actual ps output causing
> the problems that would be very helpful. It's hard to tell from the output
> exactly where the error is.
>
> I also notice that you have access to an AIX box. We haven't had (and
> still don't have) the ability to run any tests on AIX since March of this
> year because the AIX server we'd been using at PolharHome has been down
> since then. Even worse, JNA (our primary dependency) was unable to build
> new AIX binaries for the 5.10.0 release. Is there any way you can assist us
> (and JNA) with access to an AIX server for some of this testing/building?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/oshi/oshi/pull/1781#issuecomment-977064789>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOO3JCI4OMW5ZCIV4MJCWTUNPT3LANCNFSM5IPAHBVQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Rgds,
David Herschler Shvo

shvo123 commented:
Dear Daniel,

My boss approves.
What do you want me to build on AIX?
David



On Tue, Nov 23, 2021 at 9:32 PM Daniel Widdis ***@***.***>
wrote:

> Thanks for letting me know about the parsing problems. Can you please open
> a new issue for that? If possible to include the actual ps output causing
> the problems that would be very helpful. It's hard to tell from the output
> exactly where the error is.
>
> I also notice that you have access to an AIX box. We haven't had (and
> still don't have) the ability to run any tests on AIX since March of this
> year because the AIX server we'd been using at PolharHome has been down
> since then. Even worse, JNA (our primary dependency) was unable to build
> new AIX binaries for the 5.10.0 release. Is there any way you can assist us
> (and JNA) with access to an AIX server for some of this testing/building?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/oshi/oshi/pull/1781#issuecomment-977064789>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOO3JCI4OMW5ZCIV4MJCWTUNPT3LANCNFSM5IPAHBVQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Rgds,
David Herschler Shvo

shvo123 commented:
i pushed a better fix

if (m.group(4) != null) {
    milliseconds += parseLongOrDefault(m.group(4), 0L) * 1000L;
}
if (m.group(5) != null){
    milliseconds += (long) (1000 * parseDoubleOrDefault("0." + m.group(5), 0d));
}


On Wed, Nov 24, 2021 at 12:11 AM Daniel Widdis ***@***.***>
wrote:

> Did you mean to close this? :)
>
> Also while your fix works you're still spending processing time parsing a
> constant string "0". Better just to completely skip the whole += bit.
>
> Or if you want the ternary operator, do milliseconds += (m.group(5) ==
> null) ? 0 : parseDouble...
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/oshi/oshi/pull/1781#issuecomment-977215507>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOO3JCQ6BEKA3SDYJ7GIR3UNQGSRANCNFSM5IPAHBVQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Rgds,
David Herschler Shvo
